pyquil/paulis.py

```diff
@@ -324,33 +325,44 @@ def from_list(cls, terms_list, coefficient=1.0):
         pterm.coefficient = complex(coefficient)
         return pterm
 
+    @classmethod
+    def from_str(cls, str_pauli_term):
+        """Construct a PauliTerm from the result of str(pauli_term)
+        """
+        coef_str, str_pauli_term = str_pauli_term.split('*')
+        try:
+            coef = int(coef_str)
+        except ValueError:
+            try:
+                coef = float(coef_str)
+            except ValueError:
+                coef = complex(coef_str)
+
+        op = sI() * coef
+        for ma in re.finditer(r'([XYZ])(\d+)', str_pauli_term):
+            op *= PauliTerm(ma.group(1), int(ma.group(2)))
+        return op
+
     def pauli_string(self, qubits=None):
         """
-        Return a string representation of this PauliTerm mod its phase, as a concatenation of the string representation
-        of the
+        Return a string representation of this PauliTerm without its coefficient
+
         >>> p = PauliTerm("X", 0) * PauliTerm("Y", 1, 1.j)
         >>> p.pauli_string()
         "XY"
-        >>> p.pauli_string([0])
+        >>> p.pauli_string(qubits=[0])
         "X"
-        >>> p.pauli_string([0, 2])
+        >>> p.pauli_string(qubits=[0, 2])
         "XI"
 
-        :param list qubits: The list of qubits to represent, given as ints. If None, defaults to all qubits in this
-         PauliTerm.
-        :return: The string representation of this PauliTerm, modulo its phase.
-        :rtype: String
+        :param list qubits: The list of qubits to represent, given as ints. If None, defaults to
+            all qubits in this PauliTerm.
+        :return: The string representation of this PauliTerm, sans coefficient
         """
-        qubit_term_mapping = dict(self.operations_as_set())
         if qubits is None:
-            qubits = [qubit for qubit, _ in qubit_term_mapping.items()]
-        ps = ""
-        for qubit in qubits:
-            try:
-                ps += qubit_term_mapping[qubit]
-            except KeyError:
-                ps += "I"
-        return ps
+            qubits = self.get_qubits()
+
+        return ''.join(self[q] for q in qubits)
```

Comment:
This is a good change! For example, if we did `(sZ(1) * sX(0)).pauli_string()`, we'd get `'XZ'`, which disagrees with the QUIL convention of qubit ordering. With this change, we now get `'ZX'`, which agrees with the QUIL convention of qubit ordering.

Reply:
Wait. I don't want the behavior to change. Why did this cause the behavior to change?

Reply:
but qubit_term_mapping was `dict(self.operations_as_set())` which would be random order python for python < 3.6. In >= 3.6, the full call is `dict(frozenset(OrderedDict()))` which might be ordered (???) in which case it would share ordering with `self.get_qubits()`